File: android-mip-app/app/src/main/java/com/fiveq/ffci/data/api/MipApiClient.kt
package com.fiveq.ffci.data.api

import android.net.Uri
import android.util.Base64
import android.util.Log
import com.fiveq.ffci.config.AppConfig
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.coroutines.isActive
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Call
import okhttp3.logging.HttpLoggingInterceptor
import java.io.IOException
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

class ApiException(message: String, val statusCode: Int, val url: String) : Exception(message)

object MipApiClient {
    private const val TAG = "MipApiClient"
    private val resolvedPathCache = mutableMapOf<String, String>()
    
    // Load configuration from AppConfig singleton
    private val config get() = AppConfig.get()
    private val BASE_URL get() = config.apiBaseUrl
    private val API_KEY get() = config.apiKey

    private val moshi: Moshi = Moshi.Builder()
        .add(FlexibleListAdapterFactory())
        .addLast(KotlinJsonAdapterFactory())
        .build()

    private val client: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .addInterceptor(HttpLoggingInterceptor().apply {
                level = HttpLoggingInterceptor.Level.BODY
            })
            .addInterceptor { chain ->
                val credentials = "${config.username}:${config.password}"
                val basicAuth = "Basic " + Base64.encodeToString(
                    credentials.toByteArray(Charsets.UTF_8),
                    Base64.NO_WRAP
                )

                val request = chain.request().newBuilder()
                    .addHeader("Authorization", basicAuth)
                    .addHeader("X-API-Key", API_KEY)
                    .addHeader("Content-Type", "application/json")
                    .build()

                chain.proceed(request)
            }
            .build()
    }

    private suspend fun <T> fetchJson(url: String, adapter: com.squareup.moshi.JsonAdapter<T>): T {
        return withContext(Dispatchers.IO) {
            val startTime = System.currentTimeMillis()
            Log.d(TAG, "Fetching: $url")

            val request = Request.Builder()
                .url(url)
                .get()
                .build()

            try {
                client.newCall(request).execute().use { response ->
                    val duration = System.currentTimeMillis() - startTime
                    Log.d(TAG, "Response: ${response.code} in ${duration}ms")

                    if (!response.isSuccessful) {
                        throw ApiException(
                            "Failed to fetch (${response.code})",
                            response.code,
                            url
                        )
                    }

                    val body = response.body?.string()
                        ?: throw ApiException("Empty response body", response.code, url)

                    adapter.fromJson(body)
                        ?: throw ApiException("Failed to parse response", response.code, url)
                }
            } catch (e: IOException) {
                Log.e(TAG, "Network error: ${e.message}", e)
                throw e
            }
        }
    }

    suspend fun getSiteData(): SiteData {
        val adapter = moshi.adapter(SiteData::class.java)
        return fetchJson("$BASE_URL/mobile-api", adapter)
    }

    suspend fun getPage(uuid: String): PageData {
        val adapter = moshi.adapter(PageData::class.java)
        return fetchJson("$BASE_URL/mobile-api/page/$uuid", adapter)
    }

    suspend fun searchSite(query: String): List<SearchResult> {
        val encodedQuery = URLEncoder.encode(query.trim(), "UTF-8")
        val adapter = moshi.adapter<List<SearchResult>>(
            com.squareup.moshi.Types.newParameterizedType(
                List::class.java,
                SearchResult::class.java
            )
        )
        return fetchJson("$BASE_URL/mobile-api/search?q=$encodedQuery", adapter)
    }

    /**
     * Resolves a public page URL/path to a Kirby page UUID using mobile search.
     * This keeps internal-page links in-app even when HTML links are not pre-converted to /page/{uuid}.
     */
    suspend fun resolvePageUuidByUrl(url: String): String? {
        val normalizedPath = normalizePath(url) ?: return null
        resolvedPathCache[normalizedPath]?.let { return it }

        val candidateQueries = buildResolutionQueries(normalizedPath)
        if (candidateQueries.isEmpty()) return null

        for (query in candidateQueries) {
            val results = runCatching { searchSite(query) }.getOrNull() ?: continue
            val exactMatch = results.firstOrNull { result ->
                normalizePath(result.url) == normalizedPath
            }
            if (exactMatch != null) {
                resolvedPathCache[normalizedPath] = exactMatch.uuid
                return exactMatch.uuid
            }
        }

        return null
    }

    private fun normalizePath(urlOrPath: String): String? {
        val rawPath = try {
            Uri.parse(urlOrPath).path
        } catch (_: Exception) {
            null
        } ?: if (urlOrPath.startsWith("/")) {
            urlOrPath
        } else {
            return null
        }

        val trimmed = rawPath.trim()
        if (trimmed.isBlank() || trimmed == "/") return null
        return trimmed.trimEnd('/').lowercase()
    }

    private fun buildResolutionQueries(normalizedPath: String): List<String> {
        val segments = normalizedPath
            .trim('/')
            .split("/")
            .filter { it.isNotBlank() }
        if (segments.isEmpty()) return emptyList()

        val queries = linkedSetOf<String>()

        val lastSegment = segments.last()
            .replace("-", " ")
            .replace("_", " ")
            .trim()
        if (lastSegment.length >= 3) {
            queries.add(lastSegment)
        }

        val joined = segments
            .joinToString(" ")
            .replace("-", " ")
            .replace("_", " ")
            .trim()
        if (joined.length >= 3) {
            queries.add(joined)
        }

        return queries.toList()
    }
}

File: ios-mip-app/FFCI/API/MipApiClient.swift
//
//  MipApiClient.swift
//  FFCI
//
//  Firefighters for Christ International - API Client
//

import Foundation
import os.log

private let logger = Logger(subsystem: "com.fiveq.ffci", category: "API")

enum ApiError: Error {
    case invalidResponse
    case httpError(statusCode: Int, url: String)
    case decodingError(Error)
    case networkError(Error)
}

class MipApiClient {
    static let shared = MipApiClient()
    
    private let baseURL = "https://ffci.fiveq.dev"
    private let apiKey = "777359235aecc10fdfb144041dfdacfc80ca0751c7bed7b14c96f935456fc4ce"
    private let username = "fiveq"
    private let password = "demo"
    
    private let session: URLSession
    private var resolvedPathCache: [String: String] = [:]
    
    private init() {
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 30
        configuration.timeoutIntervalForResource = 30
        self.session = URLSession(configuration: configuration)
    }
    
    private func createRequest(url: URL) -> URLRequest {
        var request = URLRequest(url: url)
        
        // Basic Auth
        let credentials = "\(username):\(password)"
        if let credentialsData = credentials.data(using: .utf8) {
            let base64Credentials = credentialsData.base64EncodedString()
            request.setValue("Basic \(base64Credentials)", forHTTPHeaderField: "Authorization")
        }
        
        // API Key
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        return request
    }
    
    func getSiteData() async throws -> SiteData {
        let url = URL(string: "\(baseURL)/mobile-api")!
        logger.notice("Fetching site data from: \(url.absoluteString)")
        
        let request = createRequest(url: url)
        let startTime = Date()
        
        do {
            let (data, response) = try await session.data(for: request)
            let duration = Date().timeIntervalSince(startTime)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                throw ApiError.invalidResponse
            }
            
            logger.notice("Response: \(httpResponse.statusCode) in \(String(format: "%.0f", duration * 1000))ms")
            
            guard httpResponse.statusCode == 200 else {
                throw ApiError.httpError(statusCode: httpResponse.statusCode, url: url.absoluteString)
            }
            
            do {
                let decoder = JSONDecoder()
                let siteData = try decoder.decode(SiteData.self, from: data)
                logger.notice("Site data loaded: \(siteData.menu.count) menu items")
                return siteData
            } catch {
                logger.error("Failed to decode site data: \(error.localizedDescription)")
                throw ApiError.decodingError(error)
            }
        } catch let error as ApiError {
            throw error
        } catch {
            logger.error("Network error: \(error.localizedDescription)")
            throw ApiError.networkError(error)
        }
    }
    
    func getPage(uuid: String) async throws -> PageData {
        let url = URL(string: "\(baseURL)/mobile-api/page/\(uuid)")!
        logger.notice("Fetching page: \(uuid)")
        
        let request = createRequest(url: url)
        let startTime = Date()
        
        do {
            let (data, response) = try await session.data(for: request)
            let duration = Date().timeIntervalSince(startTime)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                throw ApiError.invalidResponse
            }
            
            logger.notice("Response: \(httpResponse.statusCode) in \(String(format: "%.0f", duration * 1000))ms")
            
            guard httpResponse.statusCode == 200 else {
                throw ApiError.httpError(statusCode: httpResponse.statusCode, url: url.absoluteString)
            }
            
            do {
                let decoder = JSONDecoder()
                let pageData = try decoder.decode(PageData.self, from: data)
                logger.notice("Page loaded: \(pageData.title), type: \(pageData.effectivePageType)")
                return pageData
            } catch {
                logger.error("Failed to decode page data: \(error.localizedDescription)")
                throw ApiError.decodingError(error)
            }
        } catch let error as ApiError {
            throw error
        } catch {
            logger.error("Network error: \(error.localizedDescription)")
            throw ApiError.networkError(error)
        }
    }
    
    func getPageWithCache(
        uuid: String,
        ttl: TimeInterval = PageCache.defaultTTL
    ) async throws -> (data: PageData, fromCache: Bool, isStale: Bool) {
        if let cached = await PageCache.shared.getAnyCache(uuid) {
            let isStale = await PageCache.shared.isExpired(uuid, ttl: ttl)
            logger.notice("Returning cached page: \(uuid, privacy: .public)")
            return (data: cached, fromCache: true, isStale: isStale)
        }
        
        let fresh = try await getPage(uuid: uuid)
        await PageCache.shared.put(uuid, data: fresh)
        return (data: fresh, fromCache: false, isStale: false)
    }
    
    func searchSite(query: String) async throws -> [SearchResult] {
        guard let encodedQuery = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) else {
            throw ApiError.networkError(NSError(domain: "Invalid query", code: -1))
        }
        
        let url = URL(string: "\(baseURL)/mobile-api/search?q=\(encodedQuery)")!
        logger.notice("Searching: \(query)")
        
        let request = createRequest(url: url)
        
        do {
            let (data, response) = try await session.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                throw ApiError.invalidResponse
            }
            
            guard httpResponse.statusCode == 200 else {
                throw ApiError.httpError(statusCode: httpResponse.statusCode, url: url.absoluteString)
            }
            
            do {
                let decoder = JSONDecoder()
                let results = try decoder.decode([SearchResult].self, from: data)
                logger.notice("Search returned \(results.count) results")
                return results
            } catch {
                logger.error("Failed to decode search results: \(error.localizedDescription)")
                throw ApiError.decodingError(error)
            }
        } catch let error as ApiError {
            throw error
        } catch {
            logger.error("Network error: \(error.localizedDescription)")
            throw ApiError.networkError(error)
        }
    }
    
    // Issue 2: Internal Non-UUID Links Resolution
    func resolvePageUuidByUrl(_ url: URL) async -> String? {
        guard let normalizedPath = normalizePath(url.absoluteString) else {
            return nil
        }

        if let cached = resolvedPathCache[normalizedPath] {
            return cached
        }

        let candidateQueries = buildResolutionQueries(normalizedPath)
        if candidateQueries.isEmpty {
            return nil
        }

        logger.notice("Resolving UUID for path: \(normalizedPath, privacy: .public)")

        for query in candidateQueries {
            guard let results = try? await searchSite(query: query) else { continue }
            if let exactMatch = results.first(where: { normalizePath($0.url) == normalizedPath }) {
                resolvedPathCache[normalizedPath] = exactMatch.uuid
                return exactMatch.uuid
            }
        }

        return nil
    }
    
    private func normalizePath(_ urlOrPath: String) -> String? {
        let rawPath: String
        if let parsedPath = URLComponents(string: urlOrPath)?.path, !parsedPath.isEmpty {
            rawPath = parsedPath
        } else if urlOrPath.hasPrefix("/") {
            rawPath = urlOrPath
        } else {
            return nil
        }

        let trimmed = rawPath.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty || trimmed == "/" {
            return nil
        }

        var normalized = trimmed.lowercased()
        while normalized.hasSuffix("/") && normalized.count > 1 {
            normalized.removeLast()
        }

        if !normalized.hasPrefix("/") {
            normalized = "/" + normalized
        }

        return normalized
    }

    private func buildResolutionQueries(_ normalizedPath: String) -> [String] {
        let segments = normalizedPath
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            .split(separator: "/")
            .map(String.init)
            .filter { !$0.isEmpty }

        if segments.isEmpty {
            return []
        }

        var queries: [String] = []

        let lastSegment = segments.last?
            .replacingOccurrences(of: "-", with: " ")
            .replacingOccurrences(of: "_", with: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if lastSegment.count >= 3 {
            queries.append(lastSegment)
        }

        let joined = segments
            .joined(separator: " ")
            .replacingOccurrences(of: "-", with: " ")
            .replacingOccurrences(of: "_", with: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        if joined.count >= 3, !queries.contains(joined) {
            queries.append(joined)
        }

        return queries
    }
}
